python_modules/dagster/dagster/core/asset_defs/asset_group.py

```diff
@@ -508,6 +513,86 @@ def materialize(
             name="in_process_materialization_job", selection=selection
         ).execute_in_process()
 
+    def prefixed(self, key_prefix: str, upstream_groups: Optional[Sequence["AssetGroup"]] = None):
```

Comment:
I think with this method it's possible to prefix an AssetGroup multiple times, e.g.:
```
group_1 = AssetGroup([asset1]).prefixed("prefix_1")
group_2 = group_1.prefixed("prefix_2")
```
and end up with an AssetKey like `AssetKey(['prefix_2', 'prefix_1', 'asset1'])`.
How will we know which asset to select for a certain prefix `AssetIn(prefix='prefix_1')`? It's unclear which asset this `AssetIn` would be referring to. Maybe `AssetIn` needs to accept a list of prefixes and match starting from the most top level prefix?

Reply:
Here's how it would work with the approach in this PR:

Let's say we've got an asset with an `AssetIn` with a prefix:
```
@asset(ins={"asset1": AssetIn(prefix="prefix1"))
def asset2(asset1):
    ...
```

That's exactly equivalent to the following:
```
@asset(ins={"asset1": AssetIn(asset_key=AssetKey(["prefix1", "asset1"])})
def asset2(asset1):
    ...
```
I.e. `asset2.dep_asset_keys == {AssetKey(["prefix1", "asset1"])}`

Now let's say we have an `AssetGroup` called `group1`  with an asset with `AssetKey(['prefix_2', 'prefix_1', 'asset1'])`:
```
assert group1.assets[0].asset_key == AssetKey(['prefix_2', 'prefix_1', 'asset1'])
```

Now we want to put `asset2` in a group and link against `asset1`:
```
group2 = AssetGroup([asset2]).prefixed("some_other_prefix", upstream_groups=[group1])
```

`prefixed` will notice that `AssetKey(["prefix1", "asset1"])` is a suffix of `AssetKey(['prefix_2', 'prefix_1', 'asset1'])` and thus match them up. If `asset2` looked like this:

```
@asset(ins={"asset1": AssetIn(prefix="prefix2"))
def asset2(asset1):
    ...
```
then there would be no match, because `AssetKey(["prefix2", "asset1"])` is not a suffix of `AssetKey(['prefix_2', 'prefix_1', 'asset1'])` 

Does that answer your question? If not, is there a situation you have in mind where there would be more ambiguity?